lib/pbench/test/unit/server/query_apis/commons.py

```diff
@@ -0,0 +1,314 @@
+
+    def test_missing_keys(self, client, server_config, user_ok, pbench_token):
+        """
+        Test behavior when JSON payload does not contain all required keys.
+
+        Note that Pbench will silently ignore any additional keys that are
+        specified but not required.
+        """
+
+        def missing_key_helper(keys):
+            response = client.post(
+                server_config.rest_uri + self.pbench_endpoint,
+                headers={"Authorization": "Bearer " + pbench_token},
+                json=keys,
+            )
+            assert response.status_code == HTTPStatus.BAD_REQUEST
+            missing = [k for k in required_keys if k not in keys]
+            assert (
+                response.json.get("message")
+                == f"Missing required parameters: {','.join(missing)}"
+            )
+
+        parameter_items = self.cls_obj.schema.parameters.items()
+
+        required_keys = [
+            key for key, parameter in parameter_items if parameter.required
+        ]
+
+        all_combinations = []
+        for r in range(1, len(parameter_items) + 1):
+            for item in itertools.combinations(parameter_items, r):
+                tmp_req_keys = [key for key, parameter in item if parameter.required]
+                if tmp_req_keys != required_keys:
+                    all_combinations.append(item)
+
+        for items in all_combinations:
+            keys = {}
+            for key, parameter in items:
+                if parameter.type == ParamType.ACCESS:
+                    keys[key] = "public"
+                elif parameter.type == ParamType.DATE:
+                    keys[key] = "2020"
+                else:
+                    keys[key] = "foobar"
+
+            missing_key_helper(keys)
+
+        # Test in case all of the required keys are missing and some
+        # random non-existent key is present in the payload
+        if required_keys:
+            missing_key_helper({"notakey": None})
+
+    def test_bad_dates(self, client, server_config, user_ok, pbench_token):
+        """
+        Test behavior when a bad date string is given
+        """
+        skip = True
+        for key, p in self.cls_obj.schema.parameters.items():
+            # Modify date/time key in the payload to make it look invalid
+            if p.type == ParamType.DATE and key in self.payload:
+                self.payload[key] = "2020-19"
+                skip = False
+        if skip:
+            pytest.skip("skipping " + self.test_bad_dates.__name__)
+        response = client.post(
+            server_config.rest_uri + self.pbench_endpoint,
+            headers={"Authorization": "Bearer " + pbench_token},
+            json=self.payload,
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert (
+            response.json.get("message")
+            == "Value '2020-19' (str) cannot be parsed as a date/time string"
+        )
```

Comment:
It'd be *so* easy now to move the call and asserts inside the loop and continue through the entire schema.

Then we could just succeed on completion; not really any need to "skip" if the API has no `DATE` parameters to test, although it might be informational.